File: Aula05/Python/Agregacao.py
#criar a classe pessoa com nome e idade
#associar uma classe endereço as pessoas onde tera alocado informações de rua, cidade, estado, cep 
#criar  uma classe empresa com nome, cnpj, e funcionarios
#fazer com que a as pessoas adicionadas sejam contratada pela empresa

class Pessoa:
    def __init__(self, nome, idade):
        self.nome = nome
        self.idade = idade
        self._endereco = None 

    def set_endereco(self, endereco):
        self._endereco = endereco

    def get_endereco(self):
        return self._endereco

    def exibir_informacoes(self):
        print(f"Nome: {self.nome} | Idade: {self.idade}")
        if self._endereco is not None:
            print("Detalhes do Endereço:")
            self._endereco.exibir_endereco()
        else:
            print("Endereço não definido.")

class Endereco:
    def __init__(self, rua, cidade, estado, cep):
        self._rua = rua
        self._cidade = cidade
        self._estado = estado
        self._cep = cep

    def exibir_endereco(self):
        print(f"Rua: {self._rua}, Cidade: {self._cidade}, Estado: {self._estado}, CEP: {self._cep}")

class Empresa:
    def __init__(self, nome, cnpj):
        self._nome = nome
        self._cnpj = cnpj
        self._funcionarios = []

    def adicionar_funcionario(self, funcionario):
        self._funcionarios.append(funcionario)

    def mostrar_funcionarios(self):
        print(f"Funcionários da empresa {self._nome}:")
        if not self._funcionarios:
            print("Nenhum funcionário contratado.")
        else:
            for funcionario in self._funcionarios:
                print(funcionario.nome)

# Exemplo de uso
end1 = Endereco("loteamento luis carlos de mendonça", "goiana", "pernambuco", "55900-000")
pessoa1 = Pessoa("otavio", 20)
pessoa1.set_endereco(end1)

end2 = Endereco("rua pal ferro", "joao pessoa", "paraiba", "58010-030")
pessoa2 = Pessoa("Fernanda", 21)
pessoa2.set_endereco(end2)

empresa = Empresa("empresa", "98765432100001")
empresa.adicionar_funcionario(pessoa1)
empresa.adicionar_funcionario(pessoa2)

empresa.mostrar_funcionarios()

File: Aula05/Go/Agregacao.go
//criar a classe pessoa com nome e idade
//associar uma classe endereço as pessoas onde tera alocado informações de rua, cidade, estado, cep
//criar  uma classe empresa com nome, cnpj, e funcionarios
//fazer com que a as pessoas adicionadas sejam contratada pela empresa

package main

import "fmt"

type Endereco struct {
	Rua    string
	Cidade string
	Estado string
	Cep    string
}

func (e *Endereco) Exibir() {
	fmt.Printf("Endereço: %s, %s - %s, CEP: %s\n", e.Rua, e.Cidade, e.Estado, e.Cep)
}

type Pessoa struct {
	Nome     string
	Idade    int
	Endereco *Endereco
}

func CriarPessoa(nome string, idade int) Pessoa {
	return Pessoa{Nome: nome, Idade: idade}
}

func (p *Pessoa) DefinirEndereco(endereco *Endereco) {
	p.Endereco = endereco
}

func (p *Pessoa) MostrarDetalhes() {
	fmt.Printf("Nome: %s, Idade: %d\n", p.Nome, p.Idade)
	if p.Endereco != nil {
		p.Endereco.Exibir()
	} else {
		fmt.Println("Endereço não informado.")
	}
}

type Empresa struct {
	Nome         string
	Cnpj         string
	Funcionarios []Pessoa
}

func CriarEmpresa(nome string, cnpj string) *Empresa {
	return &Empresa{Nome: nome, Cnpj: cnpj}
}

func (e *Empresa) AdicionarFuncionario(pessoa Pessoa) {
	e.Funcionarios = append(e.Funcionarios, pessoa)
}

func (e *Empresa) ExibirFuncionarios() {
	fmt.Printf("Funcionários da %s:\n", e.Nome)
	for _, funcionario := range e.Funcionarios {
		fmt.Println(funcionario.Nome)
	}
}

func main() {
	endereco1 := &Endereco{"Loteamento Luis Carlos de Mendonça", "Goiana", "Pernambuco", "5900-000"}
	endereco2 := &Endereco{"Rua Pau Ferro", "João Pessoa", "Paraíba", "58010-030"}

	pessoa1 := CriarPessoa("Otavio", 20)
	pessoa2 := CriarPessoa("Fernanda", 21)

	pessoa1.DefinirEndereco(endereco1)
	pessoa2.DefinirEndereco(endereco2)

	empresa := CriarEmpresa("empresa", "12345678000199")
	empresa.AdicionarFuncionario(pessoa1)
	empresa.AdicionarFuncionario(pessoa2)

	empresa.ExibirFuncionarios()
}
